src/connection.cpp

```diff
@@ -119,16 +120,53 @@ void Connection::accept()
 		readTimer.expires_from_now(boost::posix_time::seconds(CONNECTION_READ_TIMEOUT));
 		readTimer.async_wait(std::bind(&Connection::handleTimeout, std::weak_ptr<Connection>(shared_from_this()), std::placeholders::_1));
 
-		// Read size of the first packet
-		boost::asio::async_read(socket,
-		                        boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
-		                        std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
+		if (connectionState == CONNECTION_STATE_CONNECTING) {
+			// Read one-by-one ascii characters (new-line-terminated) to verify world name
+			boost::asio::async_read(socket,
+									boost::asio::buffer(msg.getBuffer(), 1),
+									std::bind(&Connection::parseProxyWorldNameIdentification, shared_from_this(), std::placeholders::_1));
+		} else {
+			// Read size of the first packet
+			boost::asio::async_read(socket,
+									boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
+									std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
+		}
+
+		
 	} catch (boost::system::system_error& e) {
 		std::cout << "[Network error - Connection::accept] " << e.what() << std::endl;
 		close(FORCE_CLOSE);
 	}
 }
 
+void Connection::parseProxyWorldNameIdentification(const boost::system::error_code& error)
+{
+	std::lock_guard<std::recursive_mutex> lockClass(connectionLock);
+	readTimer.cancel();
+
+	if (error) {
+		close(FORCE_CLOSE);
+		return;
+	} else if (connectionState == CONNECTION_STATE_DISCONNECTED) {
+		return;
+	}
+
+	char b = msg.getBuffer()[0];
+	if (b == 0x0A) {
+		if (worldName != g_config.getString(ConfigManager::SERVER_NAME)) {
+			std::cout << convertIPToString(getIP()) << " disconnected for sending invalid world name." << std::endl;
+			close();
+			return;
+		}
+
+		connectionState = CONNECTION_STATE_CONNECTED;
+	} else {
+		worldName.push_back(b);
```

Comment:
use emplace_back

Reply:
for trivial type emplace_back ? :D

Reply:
BTW if you have no idea what you are doing always use push_back.
https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/LubvTXxs7k4/ThMsbV23DQAJ